lib/services/label_expressions.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package services
+
+import (
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/gravitational/trace"
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/sirupsen/logrus"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
+
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/teleport/lib/utils/typical"
+)
+
+type labelExpression typical.Expression[labelExpressionEnv, bool]
+
+type labelExpressionEnv struct {
+	resourceLabelGetter LabelGetter
+	userTraits          map[string][]string
+}
+
+func parseLabelExpression(expr string) (labelExpression, error) {
+	if parsedExpr, ok := labelExpressionCache.Get(expr); ok {
+		return parsedExpr, nil
+	}
+	parsedExpr, err := labelExpressionParser.Parse(expr)
+	if err != nil {
+		return nil, trace.Wrap(err, "parsing label expression")
+	}
+	if evicted := labelExpressionCache.Add(expr, parsedExpr); evicted {
+		logrus.Warn("evicting entry from label expression cache")
+	}
+	return parsedExpr, nil
+
+}
+
+var (
+	labelExpressionCache  = mustNewLabelExpressionCache()
+	labelExpressionParser = mustNewLabelExpressionParser()
+)
+
+const (
+	cacheSizeEnvVar  = "TELEPORT_EXPRESSION_CACHE_SIZE"
+	defaultCacheSize = 1000
+)
+
+func mustNewLabelExpressionCache() *lru.Cache[string, labelExpression] {
+	cacheSize := defaultCacheSize
+	if env := os.Getenv(cacheSizeEnvVar); env != "" {
+		if envCacheSize, err := strconv.ParseUint(env, 10, 0); err != nil {
+			log.WithError(err).Warn("parsing " + cacheSizeEnvVar)
+		} else {
+			cacheSize = int(envCacheSize)
+		}
+	}
+	cache, err := lru.New[string, labelExpression](cacheSize)
+	if err != nil {
+		panic(err)
+	}
+	return cache
+}
+
+func mustNewLabelExpressionParser() *typical.Parser[labelExpressionEnv, bool] {
+	parser, err := typical.NewParser[labelExpressionEnv, bool](typical.ParserSpec{
+		Variables: map[string]typical.Variable{
+			"user.spec.traits": typical.DynamicVariable(
+				func(env labelExpressionEnv) (map[string][]string, error) {
+					return env.userTraits, nil
+				}),
+			"resource.metadata.labels": typical.DynamicMapFunction(
+				func(env labelExpressionEnv, key string) (string, error) {
+					label, _ := env.resourceLabelGetter.GetLabel(key)
+					return label, nil
+				}),
+			"labels": typical.DynamicMapFunction(
```

Comment:
Is "labels" an alias for "resource.metadata.labels"? Maybe use the same DynamicMapFunction instance to make it clearer?

Reply:
Thanks for raising this. It was meant to be an alias, but I had another look at `resource.metadata.labels` vs `labels` in the [resource filters docs](https://goteleport.com/docs/reference/predicate-language/#resource-filtering) and realized that that former is only the static labels while the latter is combined with the dynamic labels. I think for this we'll always want to use the combined so I've removed the `resource.metadata.labels` variable.